lib/Pingmachine/Probe/PPing.pm

```diff
@@ -0,0 +1,262 @@
+        $interval >= 1000 or
+            die "pping: calculated interval too small: $interval (step = $step, pings = $pings)\n";
+    }
+
+
+    # Make sure that we can process all hosts
+    if($interval / $hostcount < $MIN_WAIT) {
+        die "pping: step * 1000 / (pings * hostcount) must be at least 10 (step=$step, pings=$pings, hostcount=$hostcount\n";
+    }
+
+    # Prepare job
+    my %job = (
+        host2order => {},
+        output     => '',
+        pid        => '',
+    );
+    for my $order ($self->order_list->get_all) {
+        my $host = $order->pping->host;
+        push @{$job{host2order}{$host}}, $order;
+    }
+    $job{hostlist} = join("\n", shuffle keys %{$job{host2order}}) . "\n",
+    $self->current_job(\%job);
+
+    # Run pingpong
+    my $single_host = substr $job{hostlist}, 0, -1;
+    $log->info("single_host: ".$single_host);
+    my $cmd = [
+        $PPING_BIN,
+        '-id', 'pingpong_client',
+        '-mode', 'client',
+        '-interval', $interval.'ms',
+        '-count', $pings,
+        '-timeout', $TIMEOUT.'s',
+        '-format', '"%[3]s : %[5]s"',
+        '-remote', $single_host,
+        '-local', $self->source_ip,
+        '-aggregate',
+    ];
+    # Policy and performance flags
+    push @{$cmd}, $self->flags;
+    #
+    my $cmd_string = "cd $SCION_DIR && ".join(' ', @$cmd);
+    $effective_cmd = [ 'su', 'scion', '-c', $cmd_string ];
+    $log->debug("starting: @$cmd (step: $step, pings: $pings, offset: ".$self->time_offset().")") if $log->is_debug();
+
+    my $cv = run_cmd $effective_cmd,
+        '2>', '/dev/null',
+        '>', \$job{output},
+        '$$', \$job{pid};
+
+    # Install pping exit callback
+    $cv->cb(
+        sub {
+            my $cbv = shift;
+            $job{pid} = undef;
+            my $exit = $cbv->recv;
+            $exit = $exit >> 8;
+            if($exit and $exit != 1 and $exit != 2) {
+                # exit 1 means that some hosts aren't reachable
+                # exit 2 means "any IP addresses were not found"
+                $log->warning("pping seems to have failed (exit: $exit, stderr: ".$job{output}.")");
+                return;
+            }
+
+            $log->debug("finished: @$cmd (step: $step, pings: $pings, offset: ".$self->time_offset().")") if $log->is_debug();
+
+            $self->_collect_current_job();
+
+            $log->debug("collected: @$cmd (step: $step, pings: $pings, offset: ".$self->time_offset().")") if $log->is_debug();
+        }
+    );
+}
+
+sub _kill_current_job {
+    my ($self) = @_;
+
+    # Kill pping, if still running
+    my $job = $self->current_job;
+    my $job_pid = $job->{pid};
+    if($job_pid) {
+        # Check that we are killing the process we started and not an innocent bystander
+        my $cmd_match = 0;
+        if (open(proc_fh, "/proc/${job_pid}/cmdline")) {
+            $cmd_match = (join("", <proc_fh>) == "".join(' ', @$effective_cmd));
```

Comment:
This does a numeric comparison.
Better:
```
$cmd_match = (join('', readline(proc_fh)) eq join(' ', @$effective_cmd));
```
But what is $effective_cmd? A global variable! That should not be.

Reply:
I'll pass along the effective_cmd with the job, like pid is. $effective_cmd being global is indeed wrong.